reconstructSPI/iterative_refinement/expectation_maximization.py

```diff
@@ -0,0 +1,524 @@
+"""Iterative refinement with Bayesian expectation maximization."""
+
+import numpy as np
+from simSPI.transfer import eval_ctf
+
+
+class IterativeRefinement:
+    """Iterative refinement with max likelihood estimation.
+
+    Parameters
+    ----------
+    map_3d_init : arr
+        Initial particle map.
+        Shape (n_pix, n_pix, n_pix)
+    particles : arr
+        Particles to be reconstructed.
+        Shape (n_particles, n_pix, n_pix)
+    ctf_info : list of dicts
+        Each dict contains CTF k,v pairs per particle.
+            Shape (n_particles,)
+
+    References
+    ----------
+    1. Nelson, P. C. (2019). Chapter 12 : Single Particle
+    Reconstruction in Cryo-electron Microscopy.
+            In Physical Models of Living Systems (pp. 305–325).
+            https://repository.upenn.edu/physics_papers/656/
+    2. Scheres, S. H. W. (2012). RELION: Implementation of a
+    Bayesian approach to cryo-EM structure determination.
+            Journal of Structural Biology, 180(3), 519–530.
+            http://doi.org/10.1016/j.jsb.2012.09.006
+    3. Sigworth, F. J., Doerschuk, P. C., Carazo, J.-M., & Scheres,
+    S. H. W. (2010).
+            An Introduction to Maximum-Likelihood Methods in Cryo-EM.
+            In Methods in Enzymology (1st ed., Vol. 482,
+            pp. 263–294). Elsevier Inc.
+            http://doi.org/10.1016/S0076-6879(10)82011-7
+    """
+
+    def __init__(self, map_3d_init, particles, ctf_info, max_itr=7):
+        self.map_3d_init = map_3d_init
+        self.particles = particles
+        self.ctf_info = ctf_info
+        self.max_itr = max_itr
+
+    def iterative_refinement(self):
+        """Perform iterative refinement.
+
+        Acts in a Bayesian expectation maximization setting,
+        i.e. maximum a posteriori estimation.
+
+        Returns
+        -------
+        map_3d_update : arr
+            Current iteration of map.
+            Shape (n_pix, n_pix, n_pix)
+        map_3d_final : arr
+            Final updated map.
+            Shape (n_pix, n_pix, n_pix)
+        half_map_3d_final_1 : arr
+            Shape (n_pix, n_pix, n_pix)
+        half_map_3d_final_2 : arr
+            Shape (n_pix, n_pix, n_pix)
+        fsc_1d : arr
+            Final one dimensional fourier shell correlation.
+            Shape (n_pix // 2,)
+        """
+        particles_1, particles_2 = IterativeRefinement.split_array(self.particles)
+
+        ctfs = self.build_ctf_array()
+        ctfs_1, ctfs_2 = IterativeRefinement.split_array(ctfs)
+
+        # work in Fourier space. So particles can stay in Fourier
+        # space the whole time. They are experimental measurements
+        # and are fixed in the algorithm
+        particles_f_1 = IterativeRefinement.fft_3d(particles_1)
```

Comment:
eventually this will come from comSPI. 

for now I think it's fine to keep it like this, and shortly refactor after the docstrings and design is set. Sound good @fredericpoitevin  and @ninamiolane ?

Reply:
I've opened an issue https://github.com/compSPI/compSPI/issues/33